File: Python/Concurrency/Producer_consumer_demo.py
import threading
from collections import deque
from typing import Deque, Generic, TypeVar

T = TypeVar("T")


class BoundedBuffer(Generic[T]):
    def __init__(self, capacity: int) -> None:
        self._queue: Deque[T] = deque()
        self._capacity = capacity
        self._condition = threading.Condition()

    def put(self, item: T) -> None:
        with self._condition:
            while len(self._queue) == self._capacity:
                self._condition.wait()
            self._queue.append(item)
            self._condition.notify_all()

    def take(self) -> T:
        with self._condition:
            while not self._queue:
                self._condition.wait()
            item = self._queue.popleft()
            self._condition.notify_all()
            return item


def producer(buffer: "BoundedBuffer[int]") -> None:
    for i in range(10):
        buffer.put(i)
        print(f"Produced: {i}")


def consumer(buffer: "BoundedBuffer[int]") -> None:
    for _ in range(10):
        item = buffer.take()
        print(f"Consumed: {item}")


def main() -> None:
    buffer: BoundedBuffer[int] = BoundedBuffer(5)

    t1 = threading.Thread(target=producer, args=(buffer,))
    t2 = threading.Thread(target=consumer, args=(buffer,))
    t1.start()
    t2.start()
    t1.join()
    t2.join()


if __name__ == "__main__":
    main()

File: Java/Concurrency/Producer_consumer_demo.java
import java.util.LinkedList;
import java.util.Queue;

class BoundedBuffer<T> {
    private final Queue<T> queue = new LinkedList<>();
    private final int capacity;

    BoundedBuffer(int capacity) {
        this.capacity = capacity;
    }

 
    synchronized void put(T item) throws InterruptedException {
        while (queue.size() == capacity) {
            wait();
        }
        queue.add(item);

        notifyAll();
    }

    synchronized T take() throws InterruptedException {
        while (queue.isEmpty()) {
            wait();
        }

        T item = queue.poll();
        notifyAll();
        return item;
    }
}

public class Producer_consumer_demo {
    public static void main(String[] args) throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(5);

        Runnable producer = () -> {
            for (int i = 0; i < 10; i++) {
                try {
                    buffer.put(i);
                    System.out.println("Produced: " + i);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        };

        Runnable consumer = () -> {
            for (int i = 0; i < 10; i++) {
                try {
                    int item = buffer.take();
                    System.out.println("Consumed: " + item);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        };

        Thread t1 = new Thread(producer);
        Thread t2 = new Thread(consumer);
        t1.start();
        t2.start();
        t1.join();
        t2.join();
    }
}